wasm/main.py

```diff
@@ -2569,21 +2542,26 @@ def instrstarend_loop(config):
 def spec_expr(config):
     logger.debug('spec_expr()')
 
-    config["idx"] = 0
-    while 1:
-        # idx<len(instrs) since instrstar[-1]=="end" which changes instrstar
-        instruction = config["instrstar"][config["idx"]]
+
+    while config.is_executable:
+        try:
+            instruction = next(config.instructions)
+        except StopIteration:
+            break
+        #logger.debug('INSTRUCTION: %s', instruction.opcode.text)
+        #logger.debug('LOCALS: %s', config.frame.locals)
+        #logger.debug('NUM-VALUES: %d', len(config.value_stack))
+        #logger.debug('NUM-LABELS: %d', len(config.label_stack))
+        #logger.debug('NUM-FRAMES: %d', len(config.frame_stack))
+        #logger.debug('VALUES: %s', tuple(config.value_stack))
+
         logic_fn = opcode2exec[instruction.opcode][0]
-        ret = logic_fn(config)
+        logic_fn(config)
```

Comment:
This section is probably one of the larger functionality changes.  (Note that this needs to be seen in the context of the changes to `spec_end`).

Previously, execution exited once a logic function returned a truthy value.  This was a very unreliable mechanism because **any** logic function that chose to return a truthy value would have resulted in the interpreter thinking that it was done.

The new mechanism *should* correctly continue execution until it exhausts the sequence of instructions, returning whatever value is hanging out on the stack when that is done.